### PR TITLE
Changed default toa errors for event files.  can also set explicitly

### PR DESCRIPTION
**Pull request by dlakaplan**
Default errors are 1 us for all event files.  Can also set as a `Quantity` or `float`

### CODE REVIEW

**Comment by dlakaplan** [comment]
It's sort of weird to have errors associated with individual photons, but I agree they probably need to be > 0 to make basic fitting/chi^2 calculations work.

**Comment by paulray** [comment]
Looks good to me!  Thanks for the quick fix!

**Comment by paulray** [comment]
Oops, the CI doesn't like the repeated `errors` kwarg

**Comment by dlakaplan** [comment]
Yes, I fixed that.  I was actually thinking that an explicit test of this for non-Fermi satellites would also be useful (I hadn't had time/found the right test yet).  For the Fermi case it is tested.

**Comment by scottransom** [comment]
@dlakaplan The time tagging of Fermi photons is only guaranteed to about that level, though, right?  So there really is an error.  @paulray probably knows the exact number.

**Comment by scottransom** [comment]
Fermi LAT webpage says "<10us, relative to spacecraft time"
https://fermi.gsfc.nasa.gov/ssc/data/analysis/documentation/Cicerone/Cicerone_Introduction/LAT_overview.html

**Comment by paulray** [comment]
Smith et al. (2008) says: "Ground tests using cosmic ray muons demonstrated that the LAT measures event times with precision relative to UTC significantly better than a microsecond (Smith et al. 2006). On orbit, satellite
telemetry indicates comparable accuracy. The contribution to the barycentered time resolution from uncertainty in the LAT’s position is negligible."

**Comment by dlakaplan** [comment]
I guess my point was more about hard-coding a single value for all spacecraft, rather than having some uncertainty present.  But this is fine in any case.

File: src/pint/event_toas.py [hunk 1]
```diff
@@ -581,10 +585,12 @@ def load_RXTE_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf):
...
     :func:`get_RXTE_TOAs`
     """
-    return load_event_TOAs(eventname, "xte", minmjd=minmjd, maxmjd=maxmjd)
+    return load_event_TOAs(
+        eventname, "xte", minmjd=minmjd, maxmjd=maxmjd, errors=errors
+    )
 
 
-def load_NICER_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf):
+def load_NICER_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf, errors=1 * u.us):
```
Comment (paulray) [inline src/pint/event_toas.py:593]: Please make this `0.1*u.us`

Comment (dlakaplan) [reply src/pint/event_toas.py:593]: Do you want that to be across all event files?  Again, I don't know that we should put too much physical meaning here.

File: src/pint/event_toas.py [hunk 2]
```diff
@@ -550,7 +550,7 @@ def get_event_TOAs(
     )
 
 
-def load_RXTE_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf):
+def load_RXTE_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf, errors=1 * u.us):
```
Comment (paulray) [inline src/pint/event_toas.py:553]: Probably `2.5*u.us` is a better value here.

File: src/pint/event_toas.py [hunk 3]
```diff
@@ -615,10 +624,16 @@ def load_NICER_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf):
...
+        eventname,
+        "nicer",
+        minmjd=minmjd,
+        maxmjd=maxmjd,
+        errors=errors,
+    )
 
 
-def load_IXPE_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf):
+def load_IXPE_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf, errors=1 * u.us):
```
Comment (paulray) [inline src/pint/event_toas.py:636]: This one should be `20*u.us`

File: src/pint/event_toas.py [hunk 4]
```diff
@@ -649,10 +667,16 @@ def load_IXPE_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf):
...
+        eventname,
+        "ixpe",
+        minmjd=minmjd,
+        maxmjd=maxmjd,
+        errors=errors,
+    )
 
 
-def load_XMM_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf):
+def load_XMM_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf, errors=1 * u.us):
```
Comment (paulray) [inline src/pint/event_toas.py:679]: I'd use `48*u.us` for XMM

File: src/pint/event_toas.py [hunk 5]
```diff
@@ -683,10 +711,16 @@ def load_XMM_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf):
...
+        eventname,
+        "xmm",
+        minmjd=minmjd,
+        maxmjd=maxmjd,
+        errors=errors,
+    )
 
 
-def load_NuSTAR_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf):
+def load_NuSTAR_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf, errors=1 * u.us):
```
Comment (paulray) [inline src/pint/event_toas.py:723]: NuSTAR, with Matteo's calibrations, is about `65*u.us`

File: src/pint/event_toas.py (hunk 1 above)
Comment (paulray) [reply src/pint/event_toas.py:593]: Only NICER event files have 100 ns absolute accuracy to UTC.  The other missions are all worse

File: src/pint/event_toas.py [hunk 6]
```diff
@@ -717,10 +754,16 @@ def load_NuSTAR_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf):
...
+        eventname,
+        "nustar",
+        minmjd=minmjd,
+        maxmjd=maxmjd,
+        errors=errors,
+    )
 
 
-def load_Swift_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf):
+def load_Swift_TOAs(eventname, minmjd=-np.inf, maxmjd=np.inf, errors=1 * u.us):
```
Comment (paulray) [inline src/pint/event_toas.py:766]: Maybe `300*u.us` for Swift XRT data

**Comment by dlakaplan** [comment]
I made all of the uncertainties separate and set by instrument.  It would actually be better to turn the instrument-by-instrument functions into some factory functions.

**Comment by paulray** [comment]
> I made all of the uncertainties separate and set by instrument. It would actually be better to turn the instrument-by-instrument functions into some factory functions.

Looks good!  I might be better to make a factory function, but I don't think it needs to be part of this PR. I'm in favor of merging this once the tests pass.

**Comment by dlakaplan** [comment]
I can probably use `functools` to do that quickly.  Let me try while I'm on another call - I can probably do it before the tests pass

**Comment by dlakaplan** [comment]
It's pretty easy using `functools.partial`.  The only downside is that the signature changes to it's less clear what the arguments are:
```
functools.partial(<function get_event_TOAs at 0x...60d0>, mission='nicer', errors=<Quantity 0.1 us>)
    Read photon event times out of a NICER file as a :class:`pint.toa.TOAs` object
    
    Correctly handles raw event files, or ones processed with axBary to have
    barycentered TOAs. Different conditions may apply to different missions.
    
    The minmjd/maxmjd parameters can be used to avoid instantiation of TOAs
    we don't want, which can otherwise be very slow.
    
    Parameters
    ----------
    eventname : str
        File name of the FITS event list
    minmjd : float, default "-infinity"
        minimum MJD timestamp to return
    maxmjd : float, default "infinity"
        maximum MJD timestamp to return
    errors : astropy.units.Quantity or float, optional
        The uncertainty on the TOA; if it's a float it is assumed to be
        in microseconds
    ephem : str, optional
        The name of the solar system ephemeris to use; defaults to "DE421".
    planets : bool, optional
        Whether to apply Shapiro delays based on planet positions. Note that a
        long-standing TEMPO2 bug in this feature went unnoticed for years.
        Defaults to False.
```
Do you think this is OK?  The docstring can still be correct.  This is easier to update/maintain.

**Comment by dlakaplan** [comment]
Let me know if you think the latest push is OK.  It's easier than before, but you still need to set each function explicitly (more complicated things are possible but seemed harder).  It's a lot easier to read the definitions and make sure they are all consistent.  The downside is that the signature is harder to understand.

**Comment by paulray** [comment]
Yeah, that looks good. It is nice to remove all the redundancy.

**Comment by dlakaplan** [comment]
The other side of this is that the functions created with `partial` don't seem to be in the default RTD:
https://nanograv-pint--1591.org.readthedocs.build/en/1591/_autosummary/pint.event_toas.html

**Comment by paulray** [comment]
Oh, that's not optimal

**Comment by dlakaplan** [comment]
Yes, I'll see if there's a quick fix.

**Comment by dlakaplan** [comment]
Oddly, https://github.com/sphinx-doc/sphinx/pull/5305 says it will add autodocs for partial functions, and I think our sphinx should be recent enough.  But on the other hand I see people in the original issues trying with 1.8 and still having problems.  I don't know if there were divergences between the sphinx 1.8 releases (which we use) and the 2.0/3.0/... releases.  We are pinned to <2, but the current release is 7

**Comment by dlakaplan** [comment]
@paulray : please look at the lated RTD build.  I explicitly added the instrument-specific functions.  It's still not perfect, as they don't collapse like normal.  And now the instruments need to be included in 3 places.  I still feel that something like this is useful, but maybe it's better to go back a couple of pushes and deal with this other stuff later.

**Comment by paulray** [comment]
I think this looks fine. My only suggestion would be to try to get the preferred `get_XXX_TOAs` functions to appear in the docs before the `load_XXX_TOAs` functions. Is that possible?

**Comment by dlakaplan** [comment]
Changed

**Comment by dlakaplan** [comment]
It is possible to define these functions dynamically.  But I haven't found a nice way.  The best I can do is:
```
for mission in ["RXTE", "NICER", "NuSTAR", "XMM", "Swift", "IXPE"]:
    exec(
        f"get_{mission}_TOAs = partial(get_event_TOAs, mission=mission.lower(), errors=_default_uncertainty[mission])"
    )
    exec(f"get_{mission}_TOAs.__doc__ = _get_event_docstring.format(mission)")
```
which seems too messy.

**Pull request merged**